Fixed a few issues on Android + added tests

hicnar opened the pull request:
To find out what has been changed please look at the commit comments.
This PR should be followed by the removal of the fastInit option in the external API.
fixes #19
fixes #18

hicnar committed:
- added the call to PDFBoxResourceLoader.init which makes the Android code process pdf files way way faster than before
- fixed concurrency related issues in Android by removing caching of the last pdf file, and initializing of the PdfTextStripper. As a result the Dart code on an Android device may request multiple files be processed in a loop and it should not cause any troubles or app crashes like it did every time prior to the changes
- made sure the pdf doc is released/closed every time it is used (which fixes the occasional "You did not close PDF Document" warning logged by PDFBox)

hicnar committed:
- enabled multidex (otherwise tests would not work)
- added basic tests for all main methods (except scenarios with password protected pdf's).
The tests need to be run inside an emulator or real device see the comment in the @Skip annotation on very first line. Tested in Samsung Galaxy S9, Google Pixel C tablet and various iOS emulators.

AlessioLuciani commented:
@hicnar thank you for the contribution! I will look at it as soon as I can.

hicnar committed:
- fixed a few bugs mainly to do with comparing the pdf document info

AlessioLuciani commented on android/src/main/kotlin/dev/aluc/pdf_text/PdfTextPlugin.kt:
```diff
@@ -1,44 +1,31 @@
-    channel.setMethodCallHandler(PdfTextPlugin());
+    val channel = MethodChannel(flutterPluginBinding.flutterEngine.dartExecutor, "pdf_text")
+    channel.setMethodCallHandler(PdfTextPlugin())
+    PDFBoxResourceLoader.init(flutterPluginBinding.applicationContext)
```
Great! So does this essentially load the resources immediately?

hicnar replied:
Yes. And it improved the performance on Android quite a bit!

hicnar replied:
One more note here. It seems to be working fine for both pre-flutter 1.12 and post as I migrated my project literally as I was working on this stuff. It worked prior to the migration and it continuous to work now.

AlessioLuciani replied:
Perfect!

AlessioLuciani commented on android/src/main/kotlin/dev/aluc/pdf_text/PdfTextPlugin.kt:
```diff
@@ -1,44 +1,31 @@
-  private var pdfTextStripper = PDFTextStripper()
+class PdfTextPlugin: FlutterPlugin, MethodCallHandler {
 
+  @Suppress("DEPRECATION")
```
What does this deprecation refer to?

hicnar replied:
In the ```flutterPluginBinding.flutterEngine.dartExecutor``` fragment, the ```flutterEngine``` is deprecated, but I could not figure out a way to get the ```dartExecutor``` other than by doing the above

AlessioLuciani replied:
Ok, let's let it like this for now.

AlessioLuciani replied:
Nevermind, I guess we can replace that instruction with `channel = MethodChannel(flutterPluginBinding.getBinaryMessenger(), "pdf_text")` .
This should solve the problem and we can remove the suppression.
[Found here](https://stackoverflow.com/questions/61637647/flutterpluginbinding-flutterengine-is-deprecated).
@hicnar after this I think that we can merge this :D

hicnar replied:
Great find! All done now

AlessioLuciani commented on android/src/main/kotlin/dev/aluc/pdf_text/PdfTextPlugin.kt:
```diff
@@ -53,18 +40,19 @@ public class PdfTextPlugin: FlutterPlugin, MethodCallHandler {
     }
   }
 
+  @Suppress("UNCHECKED_CAST")
```
Maybe here we can check or force the cast instead of suppressing the warning.

hicnar replied:
Changed, now the unchecked_cast applies only to the ```as List<Int>``` and I'm not sure there's anything more that can be done when it comes to casting generics

AlessioLuciani commented on example/test/pdf_text_test.dart:
```diff
@@ -0,0 +1,128 @@
+@Skip('To run these tests, make sure an emulator or a real device is connected, then \'cd example\' and : \'flutter test/pfd_text_test.dart\'')
```
Are you sure that it can be launched with this command? Shouldn't "flutter" be followed by a command? I could not run the tests :( .
Great job by the way! Can't wait to see it running.

hicnar replied:
Woopsie, my bad! Absolutely you can run the tests like that! I have just somehow.. forgotten to add the 'run', so it is
```flutter run test/pdf_text_test.dart``` invoked from the example directory

hicnar replied:
Added tests for password protected pdfs. The iOS part does not work (and I haven't touched it)

AlessioLuciani replied:
Awesome, I could run the tests! Weird situation on iOS, password protected files still work for me using the example app.

hicnar replied:
@AlessioLuciani Though it should not be the case, but perhaps it is related to the way the pdf is created? I created the one the tests are run against by simply exporting it from a LibreOffice Write document on Mac. If you could create a password protected pdf (with the password = password), confirm it works in the example app when run on iOS, then just stick it in the test/resources next to the one I created, we could modify the tests to point to the one that we know works in the example app and possibly figure out what the problem is?

AlessioLuciani replied:
@hicnar now I have had a chance to try the test on iOS too (previously I had issues with running the iOS emulator and I only could on Android). I removed the if condition that stops the test for encrypted documents on iOS and run the test with your document, it seems to work for me! It does not prompt any errors and says that all tests passed. So I guess this is good news. Maybe is it that the problem is only on your emulator?

hicnar replied:
@AlessioLuciani That's really interesting! So I understand you tried to run the tests against a real device?

AlessioLuciani replied:
@hicnar no, what I described happened on my iOS emulator.

AlessioLuciani commented on lib/pdf_text.dart:
```diff
@@ -120,9 +121,12 @@ class PDFDoc {
+    /// where it begins and where it ends, therefore there is no benefit of
+    /// introducing an artificial page separator that is not part of the pdf
+    /// document per se
+    return _pages.fold<String>("", (pv, page) => "$pv${page._text}");
```
Yes, removing the "\n" makes sense. Starting the fold with "" means preceding the entire text with a "", right?

hicnar replied:
That means that the fold will start concatenating to an empty string, so essentially yes the entire text will start with an empty string.

hicnar committed:
- added tests to prove processing of encrypted pdfs works + fixed the Android part of the plugin re-enabling the processing of encrypted pdfs (the password is now passed on in all channel calls)
- TODO: the iOS implementation of the plugin does not seem to be able to process password encrypted pdfs (and I have not made any changes that could potentially be causing it)
- found some more rationale behind removing the caching of PDDocuments in Kotlin -> https://pdfbox.apache.org/2.0/faq.html, see the "Is PDFBox thread safe" section.

hicnar committed:
- unchecked cast now only applies to List<Int>
- fixed the suggestion on how to run the tests

hicnar committed:
- changed the method call returning the channel for post Flutter 1.12
